src/main/java/de/tum/cit/aet/artemis/programming/service/localci/LocalCIEventListenerService.java

```diff
@@ -0,0 +1,166 @@
+import com.hazelcast.map.listener.EntryUpdatedListener;
+
+import de.tum.cit.aet.artemis.buildagent.dto.BuildAgentInformation;
+import de.tum.cit.aet.artemis.buildagent.dto.BuildJobQueueItem;
+import de.tum.cit.aet.artemis.programming.domain.build.BuildJob;
+import de.tum.cit.aet.artemis.programming.domain.build.BuildStatus;
+import de.tum.cit.aet.artemis.programming.repository.BuildJobRepository;
+
+@Service
+@Profile("localci & scheduling")
+public class LocalCIEventListenerService {
+
+    private static final Logger log = LoggerFactory.getLogger(LocalCIEventListenerService.class);
+
+    private final HazelcastInstance hazelcastInstance;
+
+    private final LocalCIQueueWebsocketService localCIQueueWebsocketService;
+
+    private final BuildJobRepository buildJobRepository;
+
+    private final SharedQueueManagementService sharedQueueManagementService;
+
+    public LocalCIEventListenerService(@Qualifier("hazelcastInstance") HazelcastInstance hazelcastInstance, LocalCIQueueWebsocketService localCIQueueWebsocketService,
+            BuildJobRepository buildJobRepository, SharedQueueManagementService sharedQueueManagementService) {
+        this.hazelcastInstance = hazelcastInstance;
+        this.localCIQueueWebsocketService = localCIQueueWebsocketService;
+        this.buildJobRepository = buildJobRepository;
+        this.sharedQueueManagementService = sharedQueueManagementService;
+    }
+
+    /**
+     * Add listeners for build job, build agent changes.
+     */
+    @PostConstruct
+    public void init() {
+        IQueue<BuildJobQueueItem> queue = hazelcastInstance.getQueue("buildJobQueue");
+        IMap<Long, BuildJobQueueItem> processingJobs = hazelcastInstance.getMap("processingJobs");
+        IMap<String, BuildAgentInformation> buildAgentInformation = hazelcastInstance.getMap("buildAgentInformation");
+        queue.addItemListener(new QueuedBuildJobItemListener(), true);
+        processingJobs.addEntryListener(new ProcessingBuildJobItemListener(), true);
+        buildAgentInformation.addEntryListener(new BuildAgentListener(), true);
+    }
+
+    /**
+     * Check the status of pending build jobs. If a build job is missing from the queue, not being built or not finished, update the status to missing.
+     * Default interval is 5 minutes. Default delay is 1 minute.
+     */
+    @Scheduled(fixedRateString = "${artemis.continuous-integration.check-job-status-interval-seconds:300}", initialDelayString = "${artemis.continuous-integration.check-job-status-delay-seconds:60}", timeUnit = TimeUnit.SECONDS)
+    public void checkPendingBuildJobsStatus() {
+        log.info("Checking pending build jobs status");
+        List<BuildJob> pendingBuildJobs = buildJobRepository.findAllByBuildStatusIn(List.of(BuildStatus.QUEUED, BuildStatus.BUILDING));
+        for (BuildJob buildJob : pendingBuildJobs) {
+            if (buildJob.getBuildSubmissionDate().isAfter(ZonedDateTime.now().minusMinutes(5))) {
```

Comment:
I can imagine that the `ZonedDateTime.now()` call is quite expensive, isn't it? It might make sense to pull this into a variable and out of the `for` loop